src/main/java/com/amazon/corretto/hotpatch/patch/impl/set/PatchSetPatcher.java

```diff
@@ -0,0 +1,177 @@
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package com.amazon.corretto.hotpatch.patch.impl.set;
+
+import java.lang.instrument.ClassFileTransformer;
+import java.lang.instrument.Instrumentation;
+import java.lang.instrument.UnmodifiableClassException;
+import java.security.ProtectionDomain;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+import com.amazon.corretto.hotpatch.interfaces.Logger;
+import com.amazon.corretto.hotpatch.interfaces.Patcher;
+import com.amazon.corretto.hotpatch.patch.ClassTransformerHotPatch;
+
+/**
+ * A class that is capable of applying multiple {@link ClassTransformerHotPatch} together as a single patcher.
+ */
+public abstract class PatchSetPatcher implements Patcher {
+    private Instrumentation instrumentation;
+    private Logger logger;
+    private ClassFileTransformer transformer;
+
+    protected abstract List<ClassTransformerHotPatch> getPatches();
+
+    @Override
+    public int install(final Map<String, String> args, final int asmApiVersion, final Instrumentation instrumentation,
+                       final Logger logger, final boolean staticAgent) {
+        this.instrumentation = instrumentation;
+        this.logger = logger;
+
+        List<ClassTransformerHotPatch> patches = getEnabledPatches(args);
+        this.transformer = new PatchSetTransformer(patches, asmApiVersion, logger);
+
+        if (staticAgent) {
+            // As we are being installed during startup, we don't care about retransforming. The transformer will
+            // patch the classes as they are being loaded.
+            instrumentation.addTransformer(transformer);
+        } else {
+            // We are being attached to a running VM, we need to retransform the classes targeted by the patches that
+            // have already been loaded.
+            instrumentation.addTransformer(transformer, true);
+
+            boolean patchesApplied = false;
+            for (Class<?> c : instrumentation.getAllLoadedClasses()) {
+                String className = c.getName();
+                for (ClassTransformerHotPatch patch : patches) {
+                    if (patch.isTargetClass(className)) {
+                        logger.log("Patching + " + className + " (" + c.getClassLoader() + ") with patch "
+                                + patch.getName());
+                        try {
+                            instrumentation.retransformClasses(c);
+                            patchesApplied = true;
+                        } catch (UnmodifiableClassException uce) {
+                            logger.log(String.valueOf(uce));
+                        }
+                    }
+                }
+            }
+
+            if (patchesApplied) {
```

Comment:
Should be !patchesApplies, add a test for this.

Reply:
Yes it should